### PR TITLE
Add middleware for Auth0 for the Forum, add some typing to the authenticationMiddlewares file

### PR DESCRIPTION
**Pull request by jpaddison3**
The EA Forum will Soon™️ be moving to Auth0 for its authentication. This PR marks the first changes to this codebase in support of that change.

For the most part, this PR doesn't change LW functionality. The Auth0 handler will not be enabled without the database settings set. There are some changes that were caught by the additional typing I added. In particular, this should fix a bug where users who signed up via google oauth had a null email field.

I'm marking this as WIP and asking Sam to review it before handing it to y'all to review.

### Auth0 Code Description (LW reviewers feel free to skip)

For the most part, Auth0 slots in the same way as Google, Facebook, and Github. The following are the differences:

* We need to add an expressSession middleware so Auth0 can store some session state
* The Auth0 Passport strategy calls back to our userHandler with an extra argument, which we need to create a wrapper to handle
* Auth0 sometimes puts an error in req.query.error. We add some logic to handle that

Other than that, it's a laughably simple change. What was so hard again?

Sam, if you want to test, you may want to edit the Forum application to use the Username-Password-Authentication database connection instead of Forum-User-Migration, so you don't have to expose your local databases. You'll also want to add the following to line 162 of `WrappedLoginForm.tsx`

```tsx
        <a className={classes.oAuthLink} href="/auth/auth0">AUTH 0</a>
```

---

Thanks to Jim, Sam, and Ben West for working on this with me.

### CODE REVIEW

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (colophonemes) [inline packages/lesswrong/server/authenticationMiddlewares.ts:246]: We briefly discussed using something like [url-join](https://www.npmjs.com/package/url-join) for this kind of thing. Might not be worth it, but also might be good?

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (colophonemes) [inline packages/lesswrong/server/authenticationMiddlewares.ts:248]: Should this be typed with `<Auth0Profile>`?

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (colophonemes) [inline packages/lesswrong/server/authenticationMiddlewares.ts:249]: Might be nice to have a comment explaining why?

Comment (colophonemes) [reply packages/lesswrong/server/authenticationMiddlewares.ts:249]: Could also use something like lodash `omit`, I have a vague taste sense that deleting object properties is sort of poor form, but given that you're immediately returning it doesn't seem like a big deal.

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (colophonemes) [inline packages/lesswrong/server/authenticationMiddlewares.ts:256]: Is `profile.displayName` guaranteed to be set? This won't be added in the universal login, so it's not clear to me how this is going to be set immediately after login

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (colophonemes) [inline packages/lesswrong/server/authenticationMiddlewares.ts:292]: Extra space

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (colophonemes) [inline packages/lesswrong/server/authenticationMiddlewares.ts:295]: Pretty annoying that the `.authenticate()` method is so overloaded...

**Review by colophonemes — COMMENTED** [review]
Made some comments. I haven't tested or tried to break things yet, will do so and comment further if required.

**Comment by colophonemes** [comment]
I'm sure this will be fixed in a future release, but for completeness – It's pretty irritating that the `/logout` route doesn't actually log me out

**Comment by colophonemes** [comment]
Reset password looks like it works, but if I only have a Google login, then sending fails. I assume this is because the Auth0 `Forum-User-Migration` database is connected to your local machine, but might be worth just checking that all is operating as expected.

**Comment by colophonemes** [comment]
Trying to link accounts results in:
```
Error: invalid_request: the connection is not enabled
    at handleAuthenticate (/Users/sam/Sites/ea/ea-forum/Lesswrong2/packages/lesswrong/server/authenticationMiddlewares.ts:223:19)
```

Again I assume that this is related to the weirdness of the current DB setup, but FYI

**Comment by colophonemes** [comment]
If I skip account linking, I get this error:

```
Error: ID token is required but missing
    at Object.verify (/Users/sam/Sites/ea/ea-forum/Lesswrong2/node_modules/passport-auth0/lib/jwt.js:35:11)
```

**Comment by colophonemes** [comment]
If I:
- Create an account with my Google Account
- Log out
- Attempt to go through account linking but fail/skip (therefore creating another Auth0 user with the same email)
- Log in using username/password

Then it logs me in, but creates a new user account with the same email. I assume that this isn't what you want, and that this should be an error.

```
> db.users.find({email: 'sam@centreforeffectivealtruism.org'}, { username: 1, services: {auth0: {id: 1, displayName: 1 }}}).pretty()
{
        "_id" : "ei9JArtwdbjq4mxNh",
        "services" : {
                "auth0" : {
                        "displayName" : "Sam Deere",
                        "id" : "google-oauth2|110322323395817316788"
                }
        },
        "username" : "sam-deere"
}
{
        "_id" : "LiE9eteypsiSZxdsd",
        "services" : {
                "auth0" : {
                        "displayName" : "sam@centreforeffectivealtruism.org",
                        "id" : "auth0|60ad3840ca7fe6006e45380b"
                }
        },
        "username" : "sam-centreforeffectivealtruism-org"
}
```

**Comment by colophonemes** [comment]
Not sure that you should use someone's (slugified) email as the user name. This isn't immediately obvious (you only work it out if you inspect the URL of your user profile), but it seems like PII you don't want to leak (it definitely violates my expectations).

If you have to fill in this data straight away (because it's required/pre-filled in the authentication callback), I'd probably generate a random username based on the first part of the email address (e.g. `sam8439`), and then (in a future release) immediately ask the user what they'd like their display name/slug to be.

**Comment by colophonemes** [comment]
Presumably an improvement for a future time, but noticing that this part of the UI is probably not synced with Auth0, and probably should be (especially as on signup, users will be asked to verify their emails automatically with Auth0)
image

**Comment by colophonemes** [comment]
I can't get the login script to work with my local DB via PacketRiot. Get User works (I've cross-checked the id of the returned user with my local `mongo` instance), but it doesn't like it when I try to verify the password. I'm assuming that this is because I'm using the wrong password (it's the one from my password manager, but maybe it's changed, who knows ¯\_(ツ)_/¯ ). In any case, would double-check this again.

**Comment by colophonemes** [comment]
Assuming this is unrelated to anything (given the Intercom reference in the error path), but just FYI when I create a new post get about five errors like the following:

```sh
# line immediately preceding for context
2021-05-25T18:29:06.813Z: Post marked as not spam kNeCKhePALYwr9Zkp
# I get about 4 of the following:
2021-05-25T18:29:07.964Z: Unhandled rejection
2021-05-25T18:29:07.964Z: Error: User Not Found
    at callback (/Users/sam/Sites/ea/ea-forum/Lesswrong2/node_modules/intercom-client/dist/client.js:227:17)
```

File: packages/lesswrong/server/authenticationMiddlewares.ts
Comment (jimrandomh) [inline packages/lesswrong/server/authenticationMiddlewares.ts:128]: In spite of what DefinitelyTyped thinks, this is a real option; looking at the source of `passport-oauth2`, it seems to affect whether the `x-forwarded-host` http header is used. But... I have no reason to think that matters, this option probably tagged along in a copy-paste from sample code? I think it's fine to take out, but needs to be tested in the actual server environment (local testing doesn't apply). (That'll be on me to test after this merges to devel.)

**Review by jimrandomh — APPROVED** [review]

**Comment by jimrandomh** [comment]
I'll merge this into devel after the next devel-into-master, and give each of the oauth flows a test then. I don't think the risk is very high, but it can't really be tested locally.

**Comment by jpaddison3** [comment]
> I'm sure this will be fixed in a future release, but for completeness – It's pretty irritating that the /logout route doesn't actually log me out

Is already fixed in downstream branch.

**Comment by jpaddison3** [comment]
> Presumably an improvement for a future time, but noticing that this part of the UI is probably not synced with Auth0, and probably should be (especially as on signup, users will be asked to verify their emails automatically with Auth0)

Yep, tracking.

**Comment by jpaddison3** [comment]
It seems like there's something wrong with the db scripts / your setup, not sure which. Let's investigate this, but it's working for me, and I'm ok with merging but not using this code.

**Comment by jpaddison3** [comment]
> I'll merge this into devel after the next devel-into-master, and give each of the oauth flows a test then. I don't think the risk is very high, but it can't really be tested locally.

Thanks. Ready to merge on my end.

**Comment by jpaddison3** [comment]
Thanks for merging dev into this branch. I'm guessing you did some conflict-resolution — thanks.

**Comment by jpaddison3** [comment]
@jimrandomh motion to merge

**Pull request merged**
